### DISCUSSION TITLE
Don’t make the return type of `useRouteLoaderData` nullable if its argument is `"root"`

### DISCUSSION THREAD

**Discussion by lensbart** [remix-run/remix#10170] [Proposals]
If I’m not mistaken, the `root` route loader data is always available, because this route encapsulates all other routes.

It would be nice if `useRouteLoaderData` would be overloaded and would return `SerializeFrom<typeof rootLoader>` if its argument is `"root"`, and `SerializeFrom<typeof rootLoader> | undefined` (as it currently does) otherwise.

**Comment by kiliman** [comment]
I recommend creating a custom hook that you export from root
```ts
// root.tsx

export function useRootLoaderData() {
  return useLoaderData<typeof loader>()
}
```

**Reply by dvnrsn** [reply]
Is this no different from
```ts
export const useRootLoaderData = () => {
  const [root] = useMatches();
  return root?.data as SerializeFrom<typeof loader>;
};
```

**Reply by lensbart** [reply]
You need function overloads to get the suggested behavior. Easy to implement in a codebase (and we do have a `useRootLoaderData`) but the suggestion was to tighten the default types.

That being said, I think my suggestion was misguided: in case an error is thrown in the root loader, the loader data is not available in the error boundary.
